distributed/shuffle/_shuffle_extension.py

```diff
@@ -280,17 +300,23 @@ def done(self) -> bool:
         return self.transferred and self.output_partitions_left == 0
 
     async def flush_receive(self) -> None:
-        if self._exception:
-            raise self._exception
+        self.raise_if_closed()
         await self._disk_buffer.flush()
 
     async def close(self) -> None:
-        await self._comm_buffer.close()
-        await self._disk_buffer.close()
-        try:
-            self.executor.shutdown(cancel_futures=True)
-        except Exception:
-            self.executor.shutdown()
+        self.closed = True
+        async with self._close_lock:
```

Comment:
`Shuffle.close` is called in `ShuffleWorkerExtension._register_complete`,`ShuffleWorkerExtension.close` and `Shuffle.fail`, so we could see scenarios where an extension is closing and a shuffle is either completing or failing.

Fair point about idempotency. While everything should be idempotent at the moment, I'll adjust this to use the more cautious pattern of waiting on an event. Since shuffles should never reopen once closed, this should be fine.